### PR TITLE
New test to show how plugin based assets interact with BaseUrl config

### PR DESCRIPTION
**Pull request by steinkel**
Plugin based assets were not adding the plugin prefix when baseUrl was defined, for example, for working with a pull origin CDN.

### CODE REVIEW

File: src/View/Helper/UrlHelper.php [hunk 1]
```diff
@@ -176,6 +177,10 @@ public function assetUrl($path, array $options = [])
             list($plugin, $path) = $this->_View->pluginSplit($path, false);
         }
         if (!empty($options['pathPrefix']) && $path[0] !== '/') {
+            if (isset($plugin)) {
+                $path = Inflector::underscore($plugin) . '/' . $path;
```
Comment (dereuromark) [inline src/View/Helper/UrlHelper.php]: Is this really underscore, not dashed?

**Comment by steinkel** [comment]
code should pass all tests now, I'm thinking a refactor to reduce the complexity...

**Comment by steinkel** [comment]
Moving forward, there could be an explicit "CDN" configuration key, to be used as a full base when dealing with assets, and a xxxBaseUrl to deal with the relative prefix based on the type of asset, so we can build the url's using the same structure via `CDN . xxxBaseUrl . path` for assets. Fonts are a pretty common asset too, having no specific xxxBaseUrl prefix or methods in Html/Url helpers.

**Comment by markstory** [comment]
This came up once before in #10172

File: src/View/Helper/UrlHelper.php [hunk 2]
```diff
@@ -185,9 +198,10 @@ public function assetUrl($path, array $options = [])
             $path .= $options['ext'];
         }
         if (preg_match('|^([a-z0-9]+:)?//|', $path)) {
+
```
Comment (markstory) [inline src/View/Helper/UrlHelper.php]: Extra line here.

File: src/View/Helper/UrlHelper.php [hunk 3]
```diff
@@ -176,7 +177,19 @@ public function assetUrl($path, array $options = [])
             list($plugin, $path) = $this->_View->pluginSplit($path, false);
         }
         if (!empty($options['pathPrefix']) && $path[0] !== '/') {
-            $path = $options['pathPrefix'] . $path;
+            if (preg_match('|^([a-z0-9]+:)?//|', $options['pathPrefix'])) {
```
Comment (markstory) [inline src/View/Helper/UrlHelper.php]: Are you guarding against the path prefix containing a URI / full path?

**Review by markstory — COMMENTED** [review]
This has the potential to impact people's deployments as we're changing what an input path maps to in their CDN.

File: src/View/Helper/UrlHelper.php (hunk 2 above)
Comment (steinkel) [reply src/View/Helper/UrlHelper.php]: fixed

**Comment by steinkel** [comment]
Yes, reading about the context in #10172 (thanks for the link) I see others could be using a flat directory structure to copy all assets, as mentioned there.
This behavior would lead to tricky issues to debug, as cdn enabled sites usually appear later in the release cycle. As assets could overwrite other assets with the same name.

The changes implemented here pass existing tests for Html and UrlHelper, but this case was not  documented or covered in test cases, so my initial reaction was "fixing" it :)
We could:
* fix the tests and push them alone, updating the docs to clarify how the UrlHelper will process plugin based assets & absolute urls in base paths. In this case I can work on a plugin to override UrlHelper and deal with cdn's to cover these issues. This is a 100% BC scenario.
* go ahead with this change, understanding the current behavior is not correct for that specific case. In this case we could wait for a minor release, or push it to 4.x

Please let me know your thoughts, :sun_with_face:

**Comment by markstory** [comment]
Pushing this change to 4.0 might be safer. We could also use 4.x to add new/more complete configuration options to handle CDN deployments easier.

**Comment by ADmad** [comment]
I think it would be best to add a new option in 4.x.

**Comment by steinkel** [comment]
I'll rewrite the unit tests and push them alone, so others can see how the behavior works right now, thanks!

**Comment by steinkel** [comment]
unit test left only in this PR

File: tests/TestCase/View/Helper/UrlHelperTest.php [hunk 4]
```diff
@@ -544,4 +544,35 @@ public function testWebrootPaths()
...
+        $this->assertSame($expected, $result);
+
+        Configure::write('App.cssBaseUrl', $cdnPrefix);
+        $result = $this->Helper->css('TestTheme.app.css');
+        $expected = $cdnPrefix . 'app.css';
+        $this->assertSame($expected, $result);
+
+        Configure::write('App.imageBaseUrl', $imageBaseUrl);
+        Configure::write('App.jsBaseUrl', $jsBaseUrl);
+        Configure::write('App.cssBaseUrl', $cssBaseUrl);
```
Comment (markstory) [inline tests/TestCase/View/Helper/UrlHelperTest.php]: You shouldn't need to manually restore these values. The test case base class snapshots configure on each test.

Comment (steinkel) [reply tests/TestCase/View/Helper/UrlHelperTest.php]: fixed, a little too paranoid...

**Pull request merged**